server/conn.go

```diff
@@ -262,14 +264,28 @@ func (c *conn) silent() *bool {
 	return &c.silentVal
 }
 
-func (c *conn) serve(conn Conn) error {
+func (c *conn) serve(conn Conn) (err error) {
 	c.conn = conn
 
 	defer func() {
 		c.ctx.State = imap.LogoutState
 		close(c.loggedOut)
 	}()
 
+	defer func() {
+		if err := recover(); err != nil {
+			c.WriteResp(&imap.StatusResp{
+				Type: imap.StatusRespBye,
+				Info: "Internal server error, closing connection.",
+			})
+
+			stack := debug.Stack()
+			c.s.ErrorLog.Printf("panic serving %v: %v\n%s", c.Info().RemoteAddr, err, stack)
+
+			err = fmt.Errorf("%v", err)
```

Comment:
This won't set the outer error, the definition in the `if` shadows the outer one

Reply:
Oh, right.